src/io.c

```diff
@@ -1287,18 +1311,31 @@ _dispatch_fd_entry_guarded_open(dispatch_fd_entry_t fd_entry, const char *path,
 #if defined(_WIN32)
 	(void)mode;
 	DWORD dwDesiredAccess = 0;
-	if (oflag & _O_RDWR)
-		dwDesiredAccess = GENERIC_READ | GENERIC_WRITE;
-	else if (oflag & _O_RDONLY)
-		dwDesiredAccess = GENERIC_READ;
-	else if (oflag & _O_WRONLY)
-		dwDesiredAccess = GENERIC_WRITE;
+	switch (oflag & (_O_RDONLY | _O_WRONLY | _O_RDWR)) {
+		case _O_RDONLY:
+			dwDesiredAccess = GENERIC_READ;
+			break;
+		case _O_WRONLY:
+			dwDesiredAccess = GENERIC_WRITE;
+			break;
+		case _O_RDWR:
+			dwDesiredAccess = GENERIC_READ | GENERIC_WRITE;
+			break;
+	}
 	DWORD dwCreationDisposition = OPEN_EXISTING;
-	if (oflag & _O_CREAT)
+	if (oflag & _O_CREAT) {
 		dwCreationDisposition = OPEN_ALWAYS;
-	if (oflag & _O_TRUNC)
-		dwCreationDisposition = CREATE_ALWAYS;
-	return (dispatch_fd_t)CreateFile(path, dwDesiredAccess, 0, NULL, dwCreationDisposition, 0, NULL);
+		if (oflag & _O_EXCL) {
+			dwCreationDisposition = CREATE_NEW;
+		} else if (oflag & _O_TRUNC) {
+			dwCreationDisposition = CREATE_ALWAYS;
+		}
+	} else if (oflag & _O_TRUNC) {
+		dwCreationDisposition = TRUNCATE_EXISTING;
+	}
+	return (dispatch_fd_t)CreateFile(path, dwDesiredAccess,
```

Comment:
Would be nice to use `CreateFileW` so that we can handle file names which may contain non-ASCII values.

Reply:
I agree, however, the Windows port of dispatch currently does not handle Unicode well in general. I'd prefer to do a follow-up which fixes this and other similar cases.

Reply:
Sounds reasonable.